// @reduxjs/toolkit/src/index.ts
import { enableES5 } from 'immer'
export * from 'redux'
export {
  default as createNextState,
  Draft,
  current,
  freeze,
  original,
  isDraft
} from 'immer'
export {
  createSelector,
  Selector,
  OutputParametricSelector,
  OutputSelector,
  ParametricSelector
} from 'reselect'
export { createDraftSafeSelector } from '@reduxjs/toolkit/src/createDraftSafeSelector'
export { ThunkAction, ThunkDispatch } from 'redux-thunk'

// We deliberately enable Immer's ES5 support, on the grounds that
// we assume RTK will be used with React Native and other Proxy-less
// environments.  In addition, that's how Immer 4 behaved, and since
// we want to ship this in an RTK minor, we should keep the same behavior.
enableES5()

export {
  // js
  configureStore,
  // types
  ConfigureEnhancersCallback,
  ConfigureStoreOptions,
  EnhancedStore
} from '@reduxjs/toolkit/src/configureStore'
export {
  // js
  createAction,
  getType,
  // types
  PayloadAction,
  PayloadActionCreator,
  ActionCreatorWithNonInferrablePayload,
  ActionCreatorWithOptionalPayload,
  ActionCreatorWithPayload,
  ActionCreatorWithoutPayload,
  ActionCreatorWithPreparedPayload,
  PrepareAction
} from '@reduxjs/toolkit/src/createAction'
export {
  // js
  createReducer,
  // types
  Actions,
  CaseReducer,
  CaseReducers
} from '@reduxjs/toolkit/src/createReducer'
export {
  // js
  createSlice,
  // types
  CreateSliceOptions,
  Slice,
  CaseReducerActions,
  SliceCaseReducers,
  ValidateSliceCaseReducers,
  CaseReducerWithPrepare,
  SliceActionCreator
} from '@reduxjs/toolkit/src/createSlice'
export {
  // js
  createImmutableStateInvariantMiddleware,
  isImmutableDefault,
  // types
  ImmutableStateInvariantMiddlewareOptions
} from '@reduxjs/toolkit/src/immutableStateInvariantMiddleware'
export {
  // js
  createSerializableStateInvariantMiddleware,
  findNonSerializableValue,
  isPlain,
  // types
  SerializableStateInvariantMiddlewareOptions
} from '@reduxjs/toolkit/src/serializableStateInvariantMiddleware'
export {
  // js
  getDefaultMiddleware
} from '@reduxjs/toolkit/src/getDefaultMiddleware'
export {
  // types
  ActionReducerMapBuilder
} from '@reduxjs/toolkit/src/mapBuilders'
export { MiddlewareArray } from '@reduxjs/toolkit/src/utils'

export { createEntityAdapter } from '@reduxjs/toolkit/src/entities/create_adapter'
export {
  Dictionary,
  EntityState,
  EntityAdapter,
  EntitySelectors,
  EntityStateAdapter,
  EntityId,
  Update,
  IdSelector,
  Comparer
} from '@reduxjs/toolkit/src/entities/models'

export {
  AsyncThunk,
  AsyncThunkAction,
  AsyncThunkPayloadCreatorReturnValue,
  AsyncThunkPayloadCreator,
  createAsyncThunk,
  unwrapResult,
  SerializedError
} from '@reduxjs/toolkit/src/createAsyncThunk'

export {
  // js
  isAllOf,
  isAnyOf,
  isPending,
  isRejected,
  isFulfilled,
  isAsyncThunkAction,
  isRejectedWithValue,
  // types
  ActionMatchingAllOf,
  ActionMatchingAnyOf
} from '@reduxjs/toolkit/src/matchers'

export { nanoid } from '@reduxjs/toolkit/src/nanoid'

export { default as isPlainObject } from '@reduxjs/toolkit/src/isPlainObject'
